### ISSUE TITLE
Send SIGINT (Ctrl+C) on cancellation

### ISSUE THREAD

**Issue by seertenedos** [Tyrrrz/CliWrap#47]
The current cancellation token kills that application which is needed in some cases but it would also be great to have one that could close the application so the application has time to close gracefully and cleanup.  I have a few processes there it would be ideal that i could close them and only kill them as a last resort if the close does not work after a short period of time.

**Comment by Tyrrrz** [comment]
Hi @seertenedos, thank you for your issue.

I think this is a valuable idea, but it's not exactly clear to me how this should go in design-wise. As far as `Process` is concerned, we have access to [`CloseMainWindow`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.closemainwindow?view=netframework-4.8) method but it doesn't look like it takes a timeout of any sort so we may have to implement that ourselves. On top of that, I'm not sure how well it works with console applications so we'll have to test that.

Also, from a high-level API point of view in CliWrap, I tried to model CLI applications as a set of operations, in which cancellation is, I guess, expected to be instantaneous. Would it make sense if cancellation was potentially delayed by a second or two?

Generally, I'm open to this idea but there are some things that needs to be cleared out.

**Comment by seertenedos** [comment]
There is also a Close method that is not Close window. Honestly i am not sure the right one and would need to build a console app that listens to it to test i guess. As to detecting if it is sucessful maybe it could be a second cancellation token with its own method to register it for Close vs kill. That a person wants to close and then kill if it does not close fast enough they could regiser both cancelation tokens and manually fire the one to kill the process.

That way the user is in control.

**Comment by Tyrrrz** [comment]
From what I can see, `Close` method simply frees up the native handle, it doesn't actually close the application.

I guess we can extend `SetCancellationToken` with an overload to accept `bool preferGracefulTermination`.

**Comment by uAliFurkanY** [comment]
I am pretty sure that there is a close signal. Try taskkill without /F.

Like this:
```
taskkill /IM notepad.exe
SUCCESS: Sent termination signal to the process "notepad.exe" with PID 9024.
```

**Comment by Tyrrrz** [comment]
It definitely exists, but it's not provided by the `Process` class. So we'll have to have 3 native implementations for Windows/Linux/macOS (Linux and macOS might be the same).

I looked at [MedallionShell](https://github.com/madelson/MedallionShell) which claims to support sending signals, but on Window it's doing it out-of-process using an extra executable. I'm not sure why that's required but I'd guess there are some caveats to be aware of.

**Comment by uAliFurkanY** [comment]
cant you just decompile it? or spawn taskkill instead of that?

**Comment by Tyrrrz** [comment]
`taskkill` sends SIGTERM afaik, while graceful termination requires SIGINT.

Here's some info on how to send SIGINT instead: https://stackoverflow.com/a/50020028/2205454

But that will send the signal to all processes sharing the same console window which will close them all.

**Comment by uAliFurkanY** [comment]
taskkill only send sigterm when using /F

**Comment by Tyrrrz** [comment]
In that case, you're welcome to submit a PR.

**Comment by uAliFurkanY** [comment]
when I find time

**Comment by uAliFurkanY** [comment]
I think I have done it

**Comment by rgl** [comment]
Won't this work if we create the process with [CREATE_NEW_PROCESS_GROUP and then use GenerateConsoleCtrlEvent](https://docs.microsoft.com/en-us/windows/console/generateconsolectrlevent) to send a CTRL+BREAK?

**Comment by Tyrrrz** [comment]
@rgl `Process` class does not expose a way to set creation flags, so it looks like doing that would require us to completely re-implement our own process abstraction.

Also, what you're suggesting is limited to Windows. Whatever approach we come up with, it needs to be portable to Linux and Unix reasonably easy as well.

**Comment by Tyrrrz** [comment]
However, this issue on .NET repo may be of interest: https://github.com/dotnet/runtime/issues/44944

When/if they finally implement it, this approach may become viable.

**Comment by rgl** [comment]
As a side-node, the build-in Process class also does not properly work when running in a impersonation context (where we need to call the [CreateProcessAsUser](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessasuserw) win32 api). I'm testing that at https://github.com/rgl/whoami-web/tree/wip (I will also try to play with  CREATE_NEW_PROCESS_GROUP).

**Comment by chucker** [comment]
Seems related to #26, too.

https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=net-5.0#remarks itself doesn't say much on how a `CancellationToken` should behave:

>The objects that receive the notification can respond in **whatever manner is appropriate.**

https://devblogs.microsoft.com/premier-developer/recommended-patterns-for-cancellationtoken/ also doesn't seem to answer this question.

There's an added wrinkle: a command may give additional output after receiving `SIGINT`. For example, if you kill `ping` (in Unix), you get a summary/statistics. (In Windows, the default is to only ping for a certain count. In Unix, the default is to keep pinging until terminated, and then show a summary.)

Right now, I can't access whatever standard output comes after. A graceful period would help here as well.

So, at the risk of overdesigning it, I would like something like:

```csharp
.WithGracefulCancellation(trySigintFirst: true, sigkillAfter: TimeSpan.FromSeconds(5))
```

**Comment by loeffel-io** [comment]
+1

**Comment by ycherkes** [comment]
An example that shows how to do it in-proc for Windows:  [C# terminating a child process (as opposed to killing, i.e. sigkill vs. sigint on linux)](https://gist.github.com/jvshahid/6fb2f91fa7fb1db23599)

**Comment by Tyrrrz** [comment]
Yeah, I think MedallionShell's approach is the best. I'm not a fan of neither having to mess with signal handlers (MS does use a semaphore to limit access within itself, but there could be something else in the app that depends on it), nor having to deploy a separate executable to send a signal to a different console.

I am tempted to just implement the support for Linux/macOS since it's way easier and leave Windows unsupported until .NET provides a better way of doing these things.

I think the best way to implement this from the API perspective isn't to add a timeout, but to add another cancellation token. This allows the caller to offer the user an option to do both graceful and forceful termination whenever they want, instead of relying solely on time passed. Example: first click on a "cancel" button sends a signal, second click forcefully terminates the process.

I don't know yet how that would look like from design perspective, but it sounds like an interesting challenge.

**Comment by ycherkes** [comment]
@Tyrrrz 
Just a couple of thoughts, I can be wrong:
1.
```
ExecuteAsync(
	new CommandCancellation(
		gracefulCancellation: CancellationToken.None, // If this means that we aren't going to even try to stop the process
                                                              //  gracefully, then it's a little bit confusing because I would think
                                                              //  that process will be indefinitely awaited on the graceful cancellation step
		forcefulCancellation: cancellationToken
	)
)
```
2. My assumption is that more than 90% of cases will be: "just try to stop gracefully then kill it immediately". Just with probably a logging step in the mid.
3. I would like it if you keep the ability for the third-party implementation (semaphore-based or whatever) of the soft stopping of the windows process until dotnet provides a more friendly way.

**Comment by Tyrrrz** [comment]
Re 1: I see the confusion, but I think this can be clarified with documentation. There's only 4 possible states anyway, so it shouldn't be too complicated I hope.

> 2\. My assumption is that more than 90% of cases will be: "just try to stop gracefully then kill it immediately". Just with probably a logging step in the mid.

The process will not terminate immediately if stopped gracefully. In fact, there will still be a delay even if it's forcefully killed. So there needs to be an acceptable period of wait time to ensure that by a certain point we can say that it failed to terminate.


> 3\. I would like it if you keep the ability for the third-party implementation (semaphore-based or whatever) of the soft stopping of the windows process until dotnet provides a more friendly way.

You mean without the external process dance?

**Comment by ycherkes** [comment]
Re 1 and 2. OK, Can we change a paradigm and talk about a ProcessLifetimePolicy or ProcessLifetimeStrategy instead of CommandCancellation? Probably it will dismiss all the confusions.
3 Yes - please take a look at example: [Gist](https://gist.github.com/ycherkes/b216e78ae7ee4418504320f861337ed4)

**Comment by Tyrrrz** [comment]
> Re 1 and 2: OK, Can we change a paradigm and talk about a ProcessLifetimePolicy or ProcessLifetimeStrategy (or CommandLifetimePolicy, CommandLifetimeStrategy) instead of CommandCancellation? Probably it will dismiss all the confusion.

The reason I was leaning towards `CommandCancellation` was because this object is meant to replace `CancellationToken` in one of the overloads (by combining two of them), so it made sense to establish some semblance of equivalency between them:

```csharp
Command.ExecuteAsync(CancellationToken cancellationToken);
Command.ExecuteAsync(CommandCancellation commandCancellation);
```

I personally don't find that "lifetime policy" or "lifetime strategy" evoke the right set of associations here.

**Comment by ycherkes** [comment]
I meant that LifeTime isn't about cancellation only:
It can contain the whole following sentence: Let's start the process, let it work for 10 seconds, then try to stop this process gracefully, then forcibly.

**Comment by Tyrrrz** [comment]
> I meant that LifeTime isn't about cancellation only: It can contain the whole following sentence: Let's start the process, let it work for 10 seconds, then try to stop this process gracefully, then forcibly. Or even so: Let's start the process, here is the healthcheck method, please restart if the healthchek has failed.

How would that look from API perspective?

**Comment by Tyrrrz** [comment]
Some more interesting design questions:

1. In the event of graceful cancellation, should `ExecuteAsync()` throw an exception as it does for forceful cancellation? My initial instinct is to say "yes", but I'm not sure. To keep consistency with .NET guidelines, we should throw in the event of any cancellation.
2. In the event of graceful cancellation, should pipes be canceled too? I would say "no" because the process may still need them to exit gracefully (for example, by writing something useful to stdout).

**Comment by Tyrrrz** [comment]
I started a draft PR over at #179. Please discuss the implementation there if you have any comments.

**Comment by Tyrrrz** [comment]
> But you're right, if we have multiple processes- that will kill them all.
> 
> The MedallionShell has an interesting in-proc handler that prevents processes from killing [except the target one](https://github.com/madelson/MedallionShell/blob/master/MedallionShell/Signals/WindowsProcessSignaler.cs?rgh-link-date=2022-11-21T09%3A37%3A25Z#L81).

@ycherkes can you please explain how that handler prevents other processes from receiving the signal?

According to the docs, the handlers are called from last to first, until one of them returns `true`:

![image](https://user-images.githubusercontent.com/1935960/207157163-99adc3d5-affb-4aff-8cec-55ffafc5a8d5.png)

Wouldn't a race condition be possible where one of the spawned registers a handler right after us (but also right before we send the signal) and therefore receives the signal first?

I'm digging into MedallionShell's approach and the more I understand it, the less I seem to like it. I think it's just way too brittle.

**Comment by ycherkes** [comment]
@Tyrrrz Sorry, I think I confused you:

I'm not sure if there can be race conditions. It might be OK if the child process declines the CTRL+C signal by returning TRUE. That means this process can't be stopped gracefully. As I understand all Console processes have a separate thread for handling this event.

But, I noticed that:

MedallionShell works predictably with the external process signaler and CreateNoWindow = true only.
Otherwise, it stops all the processes which are sharing the same Console, except the parent.
[Example](https://gist.github.com/ycherkes/bec5f8b08f8cefdb147c6b373cb6eafa)

That's really wrong behavior. Definitely, graceful termination for Windows should be postponed.

Couple interesting links: [1 Guthub issue](https://github.com/python/cpython/issues/85323#issuecomment-1093875992), [2 Windows Internals by Russinovich](https://books.google.com.ua/books?id=y83LDgAAQBAJ&pg=PT137&lpg=PT137&dq=With+Windows+8+and+later,+the+console+architecture+changed+yet+again.+The+Conhost.exe+pro?cess+remains,+but+is+now+spawned+from+th&source=bl&ots=kIdpTwSjlB&sig=ACfU3U1lGD3aqSsx0dnANJ0KotqOfjGlXQ&hl=ru&sa=X&ved=2ahUKEwjco8nR1fb7AhVjxQIHHR1nDVMQ6AF6BAgpEAM#v=onepage&q=With%20Windows%208%20and%20later%2C%20the%20console%20architecture%20changed%20yet%20again.%20The%20Conhost.exe%20pro%3Fcess%20remains%2C%20but%20is%20now%20spawned%20from%20th&f=false)

**Comment by Tyrrrz** [comment]
> I've just figured out that handler actually doesn't pevent other processes from receiving the signal:
> 
> MedallionShell works predictably with the external process signaler and `CreateNoWindow = true` only. Otherwise, it stops all the processes which are sharing the same Console, except the signal initiator process.

Yep, that's how I understood that as well. This means, in theory, if two commands are ran simultaneously as child processes of a console window (so that they all reuse the same window), then cancelling one may also cancel the other.

Maybe it's possible to force the spawned process to create a new (hidden) console window for itself to avoid this. It seems that it can be done through Windows API using the `CREATE_NEW_CONSOLE` flag when creating a new process, but this option is not exposed through the `System.Diagnostics.Process` class. But even then, I'm not sure if there are potential drawbacks involved in this.

**Comment by ycherkes** [comment]
> Maybe it's possible to force the spawned process to create a new (hidden) console window for itself to avoid this. It seems that it can be done through Windows API using the `CREATE_NEW_CONSOLE` flag when creating a new process, but this option is not exposed through the `System.Diagnostics.Process` class. But even then, I'm not sure if there are potential drawbacks involved in this.

If we even create a process using the `CREATE_NEW_CONSOLE` flag, we still need an external process to attach to this new console. I don't see a big difference between `CreateNoWindow = true` and `CREATE_NEW_CONSOLE` in term of CTRL_C signaling. Maybe I missed something.

**Comment by Tyrrrz** [comment]
> If we even create a process using the `CREATE_NEW_CONSOLE` flag, we still need an external process to attach to this new console. I don't see a big difference between `CreateNoWindow = true` and `CREATE_NEW_CONSOLE` in term of CTRL_C signaling. Maybe I missed something.

`CreateNoWindow = true` only has an effect if the current process doesn't have a console window itself. If it already does have a console window (i.e. CliWrap is used from a console application), then the spawned process will just reuse the existing window regardless.

**Comment by Tyrrrz** [comment]
In short, these are the existing scenarios:

1. **Running inside a console application**. The spawned process will always be attached to our console window. This means that if we launch multiple processes in parallel, we can't choose which one to send Ctrl+C to (they will all get it). ❎ 
2. **Running inside a GUI application or no-UI application**. The spawned process will be attached to a new console window because our application doesn't have one. Using an external application we can selectively send Ctrl+C to just one of them. ✅ 

If we could somehow pass `CREATE_NEW_CONSOLE`, we could turn `case 1` effectively into `case 2`, if I understand how it works correctly.

**Comment by ycherkes** [comment]
> 
> `CreateNoWindow = true` only has an effect if the current process doesn't have a console window itself. If it already does have a console window (i.e. CliWrap is used from a console application), then the spawned process will just reuse the existing window regardless.

But HasSameConsole method from MedallionShell returns false for process that was 
spawned with nowindow flag from the parent console process. It means that nowindow process doesn't share the console with parent process. And that's why MedallionShell successfully sends signal via external signaler to nowindow flagged spawn. See my example above.

**Comment by Tyrrrz** [comment]
> But HasSameConsole method from MedallionShell returns false for process that was
> spawned with nowindow flag from the parent console process.

Hmmm, really? That would be surprising to me.

**Comment by ycherkes** [comment]
> Hmmm, really? That would be surprising to me.
Yup, just try an example. But it definitely should be verified on different versions of Windows. I tested on latest Win10.

**Comment by Tyrrrz** [comment]
> Yup, just try an example. But it definitely should be verified on different versions of Windows. I tested on latest Win10.

If that's how it actually works, wouldn't we be able to get away with only using the external executable approach? If the spawned processes never attach to our console, then it would always require going through the external executable approach, right?

**Comment by ycherkes** [comment]
> If that's how it actually works, wouldn't we be able to get away with only using the external executable approach? If the spawned processes never attach to our console, then it would always require going through the external executable approach, right?

Right, if CliWrap never attaches to processes that were created not by CliWrap.

**Comment by Tyrrrz** [comment]
> > If that's how it actually works, wouldn't we be able to get away with only using the external executable approach? If the spawned processes never attach to our console, then it would always require going through the external executable approach, right?
> 
> Right, if CliWrap never attaches to processes that were created not by CliWrap.

Yeah, we don't do that so that should be fine. That makes things much simpler, thanks. I was under the impression that `CreateNoWindow` still results in the process attaching to the same console window (just not creating a new one if there isn't one).

**Comment by ycherkes** [comment]
It's also interesting to check when the parent no_window process swapns other no_window process. Do they share same console? Probably no, but it must be proved to be safe.

**Comment by ycherkes** [comment]
> > 
> 
> Didn't you have something like that simulated by using the IDE test runner?

Exactly.

**Comment by ycherkes** [comment]
Success! 
If the process was created with CreateNoWindow=true, it can be gracefully terminated by an external process.
Other processes aren't affected.
I've just confirmed it in unit tests.

**Comment by Tyrrrz** [comment]
> Success!
> I've just verified it in unit tests.
> If the process was spawned with CreateNoWindow=true, it can be gracefully terminated by an external process.
> Other spawned processes didn't receive the CTRL_C signal.

Ok, awesome, thanks!

I'll finalize the changes and merge it soon.

**Comment by Tyrrrz** [comment]
Hey everyone. I just pushed a pre-release version of the package with support for graceful cancellation. I went over the API design once again and finalized it. You can find the usage instructions in the master readme: https://github.com/Tyrrrz/CliWrap#timeout-and-cancellation (scroll down to the end).

The package is here: https://www.nuget.org/packages/CliWrap/3.6.0-alpha1

Please test out the pre-release version and let me know if there are any issues. I'm planning to make a stable release next week.

**Comment by rgl** [comment]
@Tyrrrz maybe swap the arguments to make the graceful first (e.g. `ExecuteAsync(gracefulCts.Token, forcefulCts.Token);`)? I think that follows the natural flow of things (you also kinda of implied that in the order you've called `CancelAfter` at https://github.com/Tyrrrz/CliWrap#timeout-and-cancellation)

**Comment by Tyrrrz** [comment]
> @Tyrrrz maybe swap the arguments to make the graceful first (e.g. `ExecuteAsync(gracefulCts.Token, forcefulCts.Token);`)? I think that follows the natural flow of things (you also kinda of implied that in the order you've called `CancelAfter` at [Tyrrrz/CliWrap#timeout-and-cancellation](https://github.com/Tyrrrz/CliWrap#timeout-and-cancellation)). Or modify the example to call the forceful `CancelAfter` first.

This is something I was going back and forth on. The reason I settled on this order is because:
1. In the overload with a single cancellation token, that token is treated as the forceful cancellation token. It makes sense that in the overload that accepts two tokens the first token would be treated the same way.
2. I wanted to make sure that the user doesn't forget to provide the forceful token. The alternative order makes it too easy to forget about the forceful token, but you pretty much always want to pass it to make sure your command doesn't hang if the underlying process takes too long to respond to SIGINT.

**Comment by Tyrrrz** [comment]
To illustrate what I mean in `1.`, consider existing code that calls `ExecuteAsync()` like this:

```csharp
await cmd.ExecuteAsync(forcefulCancellationToken);
```

When updating that call to leverage graceful cancellation, it makes sense to me that the first parameter should have the same meaning in both overloads. Otherwise, we may end up with something like this:

```csharp
// await cmd.ExecuteAsync(forcefulCancellationToken);
await cmd.ExecuteAsync(gracefulCancellationToken, forcefulCancellationToken);
```

Which, in my opinion, is less natural. It's also a bit more error-prone, because if the first token is removed for whatever reason, the second token now has different meaning. It may be hard to catch these type of bugs because both parameters have the same type.

**Comment by rgl** [comment]
@Tyrrrz I understand the struggle, and I see/agree that forceful should be first. but please consider, always calling the argument `forcefulCancellationToken` in both examples. and in the second, initialize forceful first, e.g.:

```csharp
using var forcefulCts = new CancellationTokenSource();
using var gracefulCts = new CancellationTokenSource();

// Cancel forcefully after a timeout of 15 seconds
forcefulCts.CancelAfter(TimeSpan.FromSeconds(15));

// And also try to cancel it gracefully after a timeout of 10 seconds
// (i.e. 5 seconds before forceful cancellation) 
gracefulCts.CancelAfter(TimeSpan.FromSeconds(10));

var result = await Cli.Wrap("path/to/exe").ExecuteAsync(forcefulCts.Token, gracefulCts.Token);
```

**Comment by Tyrrrz** [comment]
> @Tyrrrz I understand the struggle, and I see/agree that forceful should be first. but please consider, always calling the argument `forcefulCancellationToken` in both examples. and in the second, initialize forceful first, e.g.:
> 
> ```cs
> using var forcefulCts = new CancellationTokenSource();
> using var gracefulCts = new CancellationTokenSource();
> 
> // Cancel forcefully after a timeout of 15 seconds
> forcefulCts.CancelAfter(TimeSpan.FromSeconds(15));
> 
> // And also try to cancel it gracefully after a timeout of 10 seconds
> // (i.e. 5 seconds before forceful cancellation) 
> gracefulCts.CancelAfter(TimeSpan.FromSeconds(10));
> 
> var result = await Cli.Wrap("path/to/exe").ExecuteAsync(forcefulCts.Token, gracefulCts.Token);
> ```

I changed the order in that example, as you suggested.

I don't want to change the other example to use `forcefulCts` instead of `cts` because I want to keep the distinction between forceful/graceful abstracted away until you actually need it (i.e., until the last example).

**Comment by rgl** [comment]
> I don't want to change the other example to use `forcefulCts` instead of `cts` because I want to keep the distinction between forceful/graceful abstracted away until you actually need it (i.e., until the last example).

not sure I understood, because if the first example is doing a forceful termination, the variable name should reflect that, and I would be explicit in its name; in my view, this would serve three purposes, make the variable intention clear for the reader, make him wonder whether there's a graceful way, and the second example would only introduce one new variable to the mix.

**Comment by Tyrrrz** [comment]
> > I don't want to change the other example to use `forcefulCts` instead of `cts` because I want to keep the distinction between forceful/graceful abstracted away until you actually need it (i.e., until the last example).
> 
> not sure I understood, because if the first example is doing a forceful termination, the variable name should reflect that, and I would be explicit in its name; in my view, this would serve three purposes, make the variable intention clear for the reader, make him wonder whether there's a graceful way, and the second example would only introduce one new variable to the mix.

I think my assumption here is the reader wouldn't even know that graceful cancellation is a thing, so they would expect the cancellation token to result in the process getting killed. Graceful cancellation is a more advanced concept, so I didn't want the first example to overload the reader with unnecessary information. At least that reflects my own background, maybe it's different depending on the person.

I will think about it in the coming days. Since it's documentation, we don't have to worry about backwards-compatibility. I'm more focused on getting the API to look right and make sure that it works for the use cases that people have. I've tested it in a few scenarios, but I easily could've missed some edge cases.

**Comment by ycherkes** [comment]
@Tyrrrz I agree with @rgl
It makes sense.
The [cancellationToken](https://github.com/Tyrrrz/CliWrap/blob/master/CliWrap/Command.Execution.cs#L293) argument should also be renamed to forcefulCancellationToken to clarify how it works without digging into the source code.
Unlike in the .NET world, when we call a method with a cancellation token, we expect the task to be canceled gracefully.

**Comment by Tyrrrz** [comment]
> Unlike in the .NET world, when we call a method with a cancellation token, we expect the task to be canceled gracefully.

Not sure what you mean here. Aren't we in .NET world?

**Comment by ycherkes** [comment]
OK, in addition to @rgl's comment :)
There is a border between .Net world of CliWrap and child process.
If this token could cross this border and inform that child process about cancellation, then it would be OK to keep it as cancellationToken.

**Comment by aborn** [comment]
https://gist.github.com/ycherkes/b216e78ae7ee4418504320f861337ed4
This is the best method to stop process gracefully.
 var success = await WindowsProcessTerminator.StopProcessGracefully(process, cts.Token);
Actually, it sends Ctrl-C to process.
Now, the problem is how to get Porcess obj. CliWrap provide processId。
Process process = Process.GetProcessById(_processId);

Thanks @ycherkes !
